File: backend/api/schemas/analyzer.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel

class ChangeRequest(BaseModel):
    # A single change request. E.g. move a gene from (old_day, old_period) to (new_day, new_period)
    # We identify the gene by class_id, day, period
    class_id: str
    subject_id: Optional[str] = None
    day: int
    period: int
    new_day: int
    new_period: int
    new_room_id: Optional[str] = None
    new_teacher_id: Optional[str] = None

class AnalyzeChangeRequest(BaseModel):
    job_id: str
    changes: List[ChangeRequest]

class RippleEffectNode(BaseModel):
    type: str # 'class', 'teacher', 'room'
    id: str
    day: int
    period: int
    clashed_with_class: Optional[str] = None

class RippleEffectResponse(BaseModel):
    direct_conflicts: List[RippleEffectNode]
    indirect_impacts: List[RippleEffectNode]

class Suggestion(BaseModel):
    day: int
    period: int
    type: str # 'move' or 'swap'
    swap_with_class_id: Optional[str] = None
    penalty: int
    penalty_delta: int
    reasons: Optional[List[str]] = None
    modified_chromosome: List[Dict[str, Any]]

class AnalyzeChangeResponse(BaseModel):
    penalty_before: int
    penalty_after: int
    penalty_delta: int
    penalty_details: Optional[Dict[str, Any]] = None
    penalty_details_before: Optional[Dict[str, Any]] = None
    ripple_effect: RippleEffectResponse
    suggestions: List[Suggestion]
    modified_chromosome: List[Dict[str, Any]]
    hill_climbed_chromosome: Optional[List[Dict[str, Any]]] = None
    hill_climb_improved: bool = False

class CommitChangeRequest(BaseModel):
    job_id: str
    new_chromosome: List[Dict[str, Any]]

class ConstraintDetail(BaseModel):
    constraint: str
    before: int
    after: int
    delta: int
    is_hard: bool
    status: str

class SubstituteTeacher(BaseModel):
    teacher_id: str
    teacher_name: str
    penalty_delta: int
    conflicts: List[str]
    constraint_details: Optional[List[ConstraintDetail]] = None
    is_qualified: bool
    is_free: bool
    modified_chromosome: List[Dict[str, Any]]

class SubstituteTeacherResponse(BaseModel):
    substitutes: List[SubstituteTeacher]

class UndoRedoRequest(BaseModel):
    job_id: str

File: frontend/src/api/analyzer.ts
import api from './client';

export interface ChangeRequest {
    class_id: string;
    subject_id?: string;
    day: number;
    period: number;
    new_day: number;
    new_period: number;
    new_room_id?: string;
    new_teacher_id?: string;
}

export interface RippleEffectNode {
    type: string;
    id: string;
    day: number;
    period: number;
    clashed_with_class?: string;
}

export interface Suggestion {
    day: number;
    period: number;
    type: string;
    swap_with_class_id: string | null;
    penalty: number;
    penalty_delta: number;
    reasons?: string[];
    modified_chromosome: any[];
}

export interface AnalyzeChangeResponse {
    penalty_before: number;
    penalty_after: number;
    penalty_delta: number;
    penalty_details?: any;
    penalty_details_before?: any;
    ripple_effect: {
        direct_conflicts: RippleEffectNode[];
        indirect_impacts: RippleEffectNode[];
    };
    suggestions: Suggestion[];
    modified_chromosome: any[];
    hill_climbed_chromosome: any[] | null;
    hill_climb_improved: boolean;
}

export interface ConstraintDetail {
    constraint: string;
    before: number;
    after: number;
    delta: number;
    is_hard: boolean;
    status: 'worsened' | 'improved';
}

export interface SubstituteTeacher {
    teacher_id: string;
    teacher_name: string;
    penalty_delta: number;
    conflicts: string[];
    constraint_details?: ConstraintDetail[];
    is_qualified: boolean;
    is_free: boolean;
    modified_chromosome: any[];
}

export interface SubstituteResponse {
    substitutes: SubstituteTeacher[];
}

export async function analyzeChange(jobId: string, changes: ChangeRequest[]): Promise<AnalyzeChangeResponse> {
    const res = await api.post(`/api/v1/analyzer/analyze-change`, { job_id: jobId, changes });
    return res.data;
}

export async function getSubstitutes(jobId: string, classId: string, day: number, period: number): Promise<SubstituteResponse> {
    const res = await api.get(`/api/v1/analyzer/substitutes`, { 
        params: { job_id: jobId, class_id: classId, day, period } 
    });
    return res.data;
}


export async function commitChange(jobId: string, newChromosome: any[]) {
    const res = await api.post(`/api/v1/analyzer/commit-change`, { job_id: jobId, new_chromosome: newChromosome });
    return res.data;
}

export async function undoChange(jobId: string) {
    const res = await api.post(`/api/v1/analyzer/undo`, { job_id: jobId });
    return res.data;
}

export async function redoChange(jobId: string) {
    const res = await api.post(`/api/v1/analyzer/redo`, { job_id: jobId });
    return res.data;
}
